test/integration/targets/cs_physical_network/tasks/main.yml

```diff
@@ -0,0 +1,205 @@
+    state: disabled
+  register: cszone
+
+- name: setup a network in check_mode
+  cs_physical_network:
+    name: net01
+    zone: "{{ cszone.name }}"
+    isolation_method: VLAN
+    broadcast_domain_range: ZONE
+  check_mode: yes
+  register: pn
+- name: validate setup a network
+  assert:
+    that:
+      - pn is successful
+      - pn is changed
+      - pn.zone == cszone.name
+
+- name: setup a network
+  cs_physical_network:
+    name: net01
+    zone: "{{ cszone.name }}"
+    isolation_method: VLAN
+    broadcast_domain_range: ZONE
+  register: pn
+- name: validate setup a network
+  assert:
+    that:
+      - pn is successful
+      - pn is changed
+      - pn.name == 'net01'
+      - pn.broadcast_domain_range == 'ZONE'
+      - pn.isolation_method == 'VLAN'
+      - pn.zone == cszone.name
+      - pn.state == 'Disabled'
+
+- name: setup a network idempotence
+  cs_physical_network:
+    name: net01
+    zone: "{{ cszone.name }}"
+    isolation_method: VLAN
+    broadcast_domain_range: ZONE
+  register: pn
+- name: validate setup a network idempotence
+  assert:
+    that:
+      - pn is successful
+      - pn is not changed
+      - pn.name == 'net01'
+      - pn.broadcast_domain_range == 'ZONE'
+      - pn.isolation_method == 'VLAN'
+      - pn.zone == cszone.name
+      - pn.state == 'Disabled'
+
+- name: set a tag on a network
+  cs_physical_network:
+    name: net01
+    tag: overlay
+    zone: "{{ cszone.name }}"
+  ignore_errors: true
+  register: pn
+- name: validate set a tag on a network
+  assert:
+    that:
+      - pn is successful
+      - pn is changed
+      - pn.name == 'net01'
+      - pn.broadcast_domain_range == 'ZONE'
+      - pn.isolation_method == 'VLAN'
+      - pn.zone == cszone.name
+      - pn.tags == 'overlay'
+      - pn.state == 'Disabled'
+
+- name: Remove tag on a network
+  cs_physical_network:
+    name: net01
+    tag: ""
+    zone: "{{ cszone.name }}"
+  register: pn
+- name: validate remove tag on a network
+  assert:
+    that:
+      - pn is successful
+      - pn is changed
+      - pn.name == 'net01'
+      - pn.broadcast_domain_range == 'ZONE'
+      - pn.isolation_method == 'VLAN'
+      - pn.zone == cszone.name
+      - pn.tags is undefined
+      - pn.state == 'Disabled'
+
+- name: ensure a network is enabled with specific nsps enabled
+  cs_physical_network:
+    name: net01
+    zone: "{{ cszone.name }}"
+    isolation_method: VLAN
+    vlan: 100-200,300-400
+    broadcast_domain_range: ZONE
+    state: enabled
+    nsps_enabled:
+      - VirtualRouter
+      - InternalLbVm
+      - VpcVirtualRouter
+  register: pn
+- name: validate ensure a network is enabled with specific nsps enabled
+  assert:
+    that:
+      - pn is successful
+      - pn is changed
+      - pn.name == 'net01'
+      - pn.broadcast_domain_range == 'ZONE'
+      - pn.isolation_method == 'VLAN'
+      - pn.zone == cszone.name
+      - pn.vlan == '100-200,300-400'
+      - pn.state == 'Enabled'
```

Comment:
Is there a way to see it enabled nsps are present?

Reply:
@dpassante currently not but I'm thinking about something like this:

```diff
diff --git a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
index 6fe7b6f54c..9699f6fccb 100644
--- a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
+++ b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
@@ -195,6 +195,52 @@ domain:
   returned: success
   type: str
   sample: domain1
+nsps:
+  description: Mapping of enabled or disabled Network Service Providers
+  type: complex
+  returned: on enabling/disabling of Network Service Providers
+  contains:
+    enabled:
+      description: Network Service Providers that were enabled
+      returned: on Network Service Provider enabling
+      type: dict
+      sample:
+      {
+        "VirtualRouter": {
+          "canenableindividualservice": true,
+          "id": "ed593eb9-3ba3-4f77-a5c9-72f28c86a156",
+          "name": "VirtualRouter",
+          "physicalnetworkid": "e54a4ee0-b1ee-4622-8f91-09048cee8315",
+          "servicelist": [
+            "Vpn",
+            "Dhcp",
+            "Dns",
+            "Gateway",
+            "Firewall",
+            "Lb",
+            "SourceNat",
+            "StaticNat",
+            "PortForwarding",
+            "UserData"
+          ],
+          "state": "Enabled"
+        }
+      }
+    disabled:
+      description: Network Service Providers that were disabled
+      returned: on Network Service Provider disabling
+      type: dict
+      sample:
+      {
+        "InternalLbVm": {
+          "canenableindividualservice": true,
+          "id": "26604aa3-ee87-4826-8572-04b7a25aaa3a",
+          "name": "InternalLbVm",
+          "physicalnetworkid": "e54a4ee0-b1ee-4622-8f91-09048cee8315",
+          "servicelist": ["Lb"],
+          "state": "Enabled"
+        }
+      }
 '''
 
 from ansible.module_utils.basic import AnsibleModule
@@ -437,14 +483,15 @@ def main():
     state = module.params.get('state')
     nsps_disabled = module.params.get('nsps_disabled', [])
     nsps_enabled = module.params.get('nsps_enabled', [])
+    nsps_result = dict()
 
     if state in ['absent']:
         network = acs_network.absent_network()
     else:
         network = acs_network.present_network()
     if nsps_disabled is not None:
-        for name in nsps_disabled:
-            acs_network.update_nsp(name=name, state='Disabled')
+        for nsp_name in nsps_disabled:
+            nsps_result['disabled'][nsp_name] = acs_network.update_nsp(name=nsp_name, state='Disabled')
 
     if nsps_enabled is not None:
         for nsp_name in nsps_enabled:
@@ -453,10 +500,11 @@ def main():
             elif nsp_name == 'InternalLbVm':
                 acs_network.set_loadbalancer_element_state(enabled=True, nsp_name=nsp_name)
 
-            acs_network.update_nsp(name=nsp_name, state='Enabled')
+            nsps_result['enabled'][nsp_name] = acs_network.update_nsp(name=nsp_name, state='Enabled')
 
     result = acs_network.get_result(network)
-
+    if nsps_result:
+        result['nsps'] = nsps_result
     module.exit_json(**result)
 
 
diff --git a/test/integration/targets/cs_physical_network/tasks/main.yml b/test/integration/targets/cs_physical_network/tasks/main.yml
index 66b8c755ab..72864a8cbf 100644
--- a/test/integration/targets/cs_physical_network/tasks/main.yml
+++ b/test/integration/targets/cs_physical_network/tasks/main.yml
@@ -126,6 +126,7 @@
       - pn.zone == cszone.name
       - pn.vlan == '100-200,300-400'
       - pn.state == 'Enabled'
+      - pn.nsps.enabled.InternalLbVm.physicalnetworkid == pn.id
 
 - name: ensure a network is disabled
   cs_physical_network:

```

Thoughts?

Reply:
It seems a bit too verbose, not knowing if all this information will be useful. However, it would be nice to have at least one list of nsps enabled / disabled.
I'm wondering if `nsps_enabled` and` nsps_disabled` only returning a list of nsps would not be sufficient.

Reply:
Yeah, I think I slightly overdid it.

This should work:

```diff
diff --git a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
index 6fe7b6f54c..1d37322bbc 100644
--- a/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
+++ b/lib/ansible/modules/cloud/cloudstack/cs_physical_network.py
@@ -195,6 +195,24 @@ domain:
   returned: success
   type: str
   sample: domain1
+nsps_enabled:
+  description: list of Network Service Providers that were enabled
+  returned: on Network Service Provider enabling
+  type: list
+  sample:
+   - VirtualRouter
+nsps_disabled:
+  description: list of Network Service Providers that were disabled
+  returned: on Network Service Provider disabling
+  type: list
+  sample:
+   - InternalLbVm
 '''
 
 from ansible.module_utils.basic import AnsibleModule
@@ -452,11 +471,13 @@ def main():
                 acs_network.set_vrouter_element_state(enabled=True, nsp_name=nsp_name)
             elif nsp_name == 'InternalLbVm':
                 acs_network.set_loadbalancer_element_state(enabled=True, nsp_name=nsp_name)

             acs_network.update_nsp(name=nsp_name, state='Enabled')
 

     result = acs_network.get_result(network)
+    result['nsps_enabled'] = nsps_enabled
+    result['nsps_disabled'] = nsps_disabled

     module.exit_json(**result)
 
 
diff --git a/test/integration/targets/cs_physical_network/tasks/main.yml b/test/integration/targets/cs_physical_network/tasks/main.yml
index 66b8c755ab..5394963b31 100644
--- a/test/integration/targets/cs_physical_network/tasks/main.yml
+++ b/test/integration/targets/cs_physical_network/tasks/main.yml
@@ -126,6 +126,7 @@
       - pn.zone == cszone.name
       - pn.vlan == '100-200,300-400'
       - pn.state == 'Enabled'
+      - 'InternalLbVm' in pn.nsps_enabled
 
 - name: ensure a network is disabled
   cs_physical_network:
```

Reply:
LGTM

Reply:
Done.